server/WFCORE-6750-unstable-api-annotation-scanner.adoc

```diff
@@ -0,0 +1,243 @@
+**** If the attribute is set to `error`, the deployment will fail with an exception containing the usages.
+
+
+
+=== Nice-to-Have Requirements
+// Requirements in this section do not have to be met to merge the proposed functionality.
+// Note: Nice-to-have requirements that don't end up being implemented as part of
+// the work covered by this proposal should be moved to the 'Future Work' section.
+
+
+=== Non-Requirements
+// Use this section to explicitly discuss things that readers might think are required
+// but which are not required.
+* Full indexing of all annotations indicating unstable API
+** See the 'Future Work' section for plans to get full coverage
+* Inspection of user classes not in an application deployment, such as:
+** User provided modules
+** Classes provided by the user via the management model with the typical `module`/`class` attribute pairs.
+* No attempt is made to not add the indices if the server is provisioned a a higher stability level than 'preview'. In other words, if the feature packs contain the indices, they will always end up in the server.
+* Since this is a `preview` feature, and the model version is the same as the original with the added `preview` qualifier, model transformation is not needed.
+
+=== Future Work
+// Use this section to discuss requirements that are not addressed by this proposal
+// but which may be addressed in later proposals.
+* Promote this feature to Community or higher.
+** This would allow us to enable it out of the box by default, making it more useful.
+* Identify all annotations indicating unstable API used by WildFly itself with the help of component leads.
+** Scan for these annotations during feature pack builds, and make them available in a provisioned server.
+* Engage with owners of feature packs external to WildFly, and identify all annotations indicating unstable API in those feature packs.
+** Provide instructions on how to do the build time scan, and to make the indices available at runtime once their feature pack has been provisioned.
+** Scan for these annotations during feature pack build, and make them available in a provisioned server.
+** I think an appropriate set of feature packs would be the ones WildFly Glow is aware of.
+* Index user provided classes
+** in user modules
+** in classes referenced via the `module`/`class` attribute pairs in the management model
+* Add a deployment marker to force enable/disable the runtime scanning independent of the subsystem setting
+** This could possibly be enhanced with a filter so a user could say to ignore usage of constructs annotated with `org.hibernate.Incubating` (or perhaps an expression to ignore classes matching the expression), while reporting usage of everything else (e.g. constructs annotated with `io.smallrye.common.annotation.Experimental`).
```

Comment:
If we're going to allow something like this, which involves the user identifying annotations of interest, then having that config do a mapping to a stability level seems good. Then how we treat the presence of the thing depends on the server's stability level.

If we found something that wasn't configured, that would mean 'not allowed' (so log it or error). IOW, not configured != experimental. Someone running a --stability=experimental server can still reject APIs.

Reply:
I consider this more brainstorming, than a "must" have. I've updated the start of the future work section to say we will decide the scope when going to community level, and that these points are ideas

Reply:
Great; that was my intent. :)